### ISSUE TITLE
Should override errors for abstract mixed-in declarations be omitted?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#1288]
Consider the following library:

```dart
mixin M1 {
  void m();
}

abstract class A {
  void m();
}

mixin M2 implements A {}

class C {
  void m([arg]) {}
}

class D1 = C with M1; // Error.
class D2 = C with A; // Error.
class D3 = C with M2; // OK.
```

The code illustrates that if we have a mixin like `M1` with an abstract declaration, we get a compile-time error when that mixin is mixed-in with a class where the interface already has a more specific signature for the given member.

Concretely, `void m()` is an incorrect override error when `M1` is mixed-in with `C` because `C` has the member signature `void m([dynamic])` already.

The mixin derived from `A` works similarly.

However, the mixin `M2` and class `D3` illustrate that this error can be eliminated by moving the offending declaration from the mixin into a superinterface.

Note that the error is not required, because an abstract declaration does not participate in overriding. It is sufficient to treat it the same way as a member signature which is contributed by a superinterface. We will still get errors for conflicting member signatures, we're just allowing an abstract declaration in a mixin to be less demanding than the superclass that it will be mixed-in with.

Hence, we could omit the override error for abstract declarations, such that applying both `M1` or `M2` to `C` would work.

The work-around (of moving the abstract declaration into a superinterface) is of course available, but it should be applied to the mixin in order to allow an unknown and extensible set of clients to use it. This favors a style where mixins never have any abstract members: Abstract declarations should always be moved to some superinterface (e.g., an abstract class which was created just for this purpose), such that no mixin application will have any of these spurious override violation errors.

We might as well eliminate those errors proactively in the language, and thus allow mixin authors to freely add abstract declarations to mixins as needed.

@munificent, @lrhn, @natebosch, @jakemac53, @leafpetersen, @stereotype441, WDYT?

**Comment by lrhn** [comment]
I think it's wrong that there is an error to begin with, but possibly it's under under- or mis-specified why.

The class `C with M` implements both `C` and `M` and adds the declarations of `M` as members of `C with M`. *That should probably only be the concrete members of `M`*.

The `C with M` already implements `M`, so it's redundant to also treat abstract declarations of `M` as declarations of `C with M`. It's doing that which causes the error here. The declaration `void m();` is an invalid override of `void m([arg]);`, but but if `void m();` comes from a superinterface of `M` then there is no declaration moved into `C with M`, and we just use our normal combined superinterface algorithm to find the most specific member signature (which is the one from `C`).

I think it's fine to have an error if you declare a too specific signature in a subclass, but I don't think it's worth it to make it an error to get an error from an abstract declaration in a mixin. That one should not be "mixed in" at all.

**Comment by lrhn** [comment]
I think it's wrong that there is an error to begin with, but possibly it's under under- or mis-specified why.

The class `C with M` implements both interfaces `C` and `M` and adds the declarations of `M` as members of `C with M`. That should probably only be the *concrete* member declarations of `M`.

The `C with M` already implements `M`, so it's redundant to also treat abstract declarations of `M` as declarations of `C with M`. It's doing that which causes the error here. The declaration `void m();` is an invalid override of `void m([arg]);`, but but if `void m();` comes from a superinterface of `M` then there is no declaration moved into `C with M`, and we just use our normal combined superinterface algorithm to find the most specific member signature (which is the one from `C`).

I think it's fine to have an error if you declare a too specific signature in a subclass, it's *always* useless, but I don't think it's worth it to make it an error to get an error from an abstract declaration in a mixin. That one should not be "mixed in" at all, it's just a requirement that the final class should implement somehow, not an *override* of a concrete super-declaration.

**Comment by eernstg** [comment]
> should probably only be the _concrete_ member declarations of `M`.

Yes, I also considered using that description of the desired outcome.

> fine to have an error if you declare a too specific signature in a subclass, it's _always_ useless

Right, and even misleading for a reader of the code. We could make it a warning, though, a bit like dead code.

We could then say that abstract declarations are never subject to override checks, they only contribute to the set of member signatures considered during computations of interfaces.

**Comment by lrhn** [comment]
I don't like warnings for something which is useless. Just make it an error and make the user fix it. (Then, possibly, provide a way to compile programs with errors if you don't care about fixing it *right now* while developing, but it's not something your final program should contain).

**Comment by eernstg** [comment]
Here is a comment from @stereotype441 on a duplicate of this issue, #1287, copied here when #1287 was closed:

**Comment by Levi-Lesches** [comment]
> Just make it an error and make the user fix it. (Then, possibly, provide a way to compile programs with errors if you don't care about fixing it _right now_ while developing, but it's not something your final program should contain).

Isn't this just a warning though?
